nixos/modules/services/networking/frr.nix

```diff
@@ -0,0 +1,186 @@
+          !
+          ${scfg.config}
+          !
+          end
+        '';
+
+  serviceOptions = service:
+    {
+      enable = mkEnableOption "the FRR ${toUpper service} routing protocol";
+
+      configFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        example = "/etc/frr/${daemonName service}.conf";
+        description = ''
+          Configuration file to use for FRR ${daemonName service}.
+          By default the NixOS generated files are used.
+        '';
+      };
+
+      config = mkOption {
+        type = types.lines;
+        default = "";
+        example =
+          let
+            examples = {
+              rip = ''
+                router rip
+                  network 10.0.0.0/8
+              '';
+
+              ospf = ''
+                router ospf
+                  network 10.0.0.0/8 area 0
+              '';
+
+              bgp = ''
+                router bgp 65001
+                  neighbor 10.0.0.1 remote-as 65001
+              '';
+            };
+          in
+            examples.${service} or "";
+        description = ''
+          ${daemonName service} configuration statements.
+        '';
+      };
+
+      vtyListenAddress = mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        description = ''
+          Address to bind to for the VTY interface.
+        '';
+      };
+
+      vtyListenPort = mkOption {
+        type = types.nullOr types.int;
+        default = null;
+        description = ''
+          TCP Port to bind to for the VTY interface.
+        '';
+      };
+    };
+
+in
+
+{
+
+  ###### interface
+  imports = [
+    {
+      options.services.frr = {
+        zebra = (serviceOptions "zebra") // {
+          enable = mkOption {
+            type = types.bool;
+            default = any isEnabled services;
+            description = ''
+              Whether to enable the Zebra routing manager.
+
+              The Zebra routing manager is automatically enabled
+              if any routing protocols are configured.
+            '';
+          };
+        };
+      };
+    }
+    { options.services.frr = (genAttrs services serviceOptions); }
+  ];
+
+  ###### implementation
+
+  config = mkIf (any isEnabled allServices) {
+
+    environment.systemPackages = [
+      pkgs.frr               # for the vtysh tool
+    ];
+
+    users.users.frr = {
+      description = "FRR daemon user";
+      isSystemUser = true;
+      group = "frr";
+    };
+
+    users.groups = {
+      frr = {};
+      # Members of the frrvty group can use vtysh to inspect the FRR daemons
+      frrvty = { members = [ "frr" ]; };
+    };
+
+    systemd.services =
+      let
+        frrService = service:
+          let
+            scfg = cfg.${service};
+            daemon = daemonName service;
+          in
+            nameValuePair daemon ({
+              wantedBy = [ "multi-user.target" ];
+              restartTriggers = [ (configFile service) ];
```

Comment:
This is redundant and can be removed.

`restartTriggers` doesn't actually **do** anything... the intention of it is to introduce text which will be different every time you rebuild (because the path in the nix store is different) - this causes the service to be restarted because the text in the service changed. You already reference the configuration file elsewhere in the service file.

Reply:
Though maybe you want to have it reloaded, not restarted should the configuration change.

That's what `reloadIfChanged` is for.

Reply:
Though if you enable `reloadIfChanged` your service won't restart when the binary changes... potentially introducing security vulnerabilities. It's a delicate balance on when to use that or not.

Regardless though, `restartTriggers` is not needed.

Reply:
Routing daemons are really stateful beasts. If handled wrong they interrupt production traffic, that's usually not a great spot to be in.

Reply:
I'll be honest... I've been really busy and didn't even read what this software is... I figured you had it covered, but just wanted to mention the `restartTriggers` bit as most people don't understand the option (it's very complicated and "broken").

`reloadIfChanged` does sound good in that case!

:smile:

Reply:
I'll keep it default `restartIfChanged` for now, because this is what quagga was doing and I haven't heard of problems. Will remove restartTriggers, good point.